### ISSUE TITLE
Principles for deployment

### ISSUE THREAD

**Issue by rhoadesre** [isaqb-org/curriculum-foundation#259]
It would be helpful to describe specific principles that would improve the implementation for deployment. These should not be implementation-specific (e.g. no frameworks, products, libraries etc.), but more general principles.

**Comment by rhoadesre** [comment]
@rhoadesre please provide example LG formulation for it...
please add quotable references

**Comment by rhoadesre** [comment]
Here are some examples:

### Release IT
**Michael T. Nygard**

- Automate Deployments
- Use immutable and disposable infrastructure
- Continuously deploy
- Deploy in phases (e.g., database, web assets, code, cleanup)
- Manage versions
- ...

### Software Deployment Principles & Practices
**Thyagarajan Krishnan**
https://www.slideshare.net/thyagk/software-deployment-principles-practices

**Repeatable Builds**
- Versioning of all software artefacts, tools, and dependencies
- Baselining all artefacts together
- Automated and documented build process

**Consistent environments**
- The environment includes all the components needed to build and run the application, including, for example, operating system, system libraries, runtime environments, application servers, etc.
- All environments (development, staging, etc.) should use the same components as the production environment
- Version control use used for all components
- Storage of the components in centralized
- A strategy exists for running multiple versions of components
- The deployment of components is automated
- All application dependencies are identified and documented
- A virtualization strategy is specified
- A change management tool is used to track environment changes

**Autonomous Packages**
- Deployable packages are independent
- Multiple instances can be executed simultaneously
- Relatve paths are used for configuration, logs, etc.
- Application install, uninstall, start, and stop processes or automation is specific to each independent package
- Each deployment package contains its own dependencies

**Easy to undo releases**
- Software releases can be installed, uninstalled or rolled back, e.g. with documented release strategy, dependency management, automated releases, backups etc.

### Software Deployment in 2022: Deployment Strategies and Best Practices
https://codefresh.io/learn/software-deployment/

**Deployment Strategies**

- Basic, rolling, multi-service, blue/green, canary, A/B testing, shadow

**Deployment best practices**

- Separate Clusters for Production and Non-Production
- Application of resource limits
- Use of deployment metrics
- Fully automated rollbacks
- "Secrets" strategy
- 

### The 9 Principles of Continuous Delivery
https://www.dbmaestro.com/blog/database-ci-cd/principles-continuous-delivery

- Repeatable Reliable Process
- Automate Everything
- Version Control Everything
- Bring the Pain Forward
- Build-in Quality
- "Done" Means Released
- Everyone is Responsible
- Continuous Improvement
- Don’t Forget the Database

### 12 Factor App
(more than just deployment)
https://12factor.net/

I. Codebase
	One codebase tracked in revision control, many deploys
II. Dependencies
	Explicitly declare and isolate dependencies
III. Config
	Store config in the environment
IV. Backing services
	Treat backing services as attached resources
V. Build, release, run
	Strictly separate build and run stages
VI. Processes
	Execute the app as one or more stateless processes
VII. Port binding
	Export services via port binding
VIII. Concurrency
	Scale out via the process model
IX. Disposability
	Maximize robustness with fast startup and graceful shutdown
X. Dev/prod parity
	Keep development, staging, and production as similar as possible
XI. Logs
	Treat logs as event streams
XII. Admin processes
	Run admin/management tasks as one-off processes

**Comment by gernotstarke** [comment]
thanx, now I know what you originally meant with your proposal.

+1 from me.

**Comment by gernotstarke** [comment]
either

(1) amend and re-formulate LG 2-4 (crosscutting concepts)
(2) add a new LG 2-10 „Know and Appropriately Apply Principles of Software Deployment“

We vote within the FLWG

**Comment by rhoadesre** [comment]
Option (3) amendment the  “LG 2-6: Explain and use design principles”.

**Comment by p-goetz** [comment]
I very much like the proposal and the list of principles and examples and have added a few of those to my "To Read" list. ;)
I have a clarifying question, though: should this list be added "as is" to the curriculum? Then I think the sheer amount of content might decrease the visibility of R1 and other R2 content. It would feel like we add a lot of room in the curriculum to a topic of less relevance (R2/R3). 
But maybe I didn't get the idea and am happy for an explanation that destroys my doubt.

**Comment by gernotstarke** [comment]
we will add a (very) condensed version, approx 5-8 items in a bullet list

**Comment by ulibecker** [comment]
These are important topics, but I am not convinced there is enough room in the foundation level to handle this appropriately. Just having an additional topic that is handled very briefly does not make the curriculum better IMO. Another aspect is that the relevance of these principles depends a lot on the type of system. 

If we include it at all, I prefer the separate learning goal, and only as R3.

**Comment by alxlo** [comment]
From a topic perspective it fits rather good in LG 2-4, but we shouldn't have a very short R1 general part on cross-cutting concepts and a equally large R3 part on one specific concept (and no other specific concepts).

But if we keep it **very** brief (maybe 2-3 bullet points with references to details) we should put it as R3 in LG 2-4. This gives us the option to do likewise when we stumble upon another important concept in the future (maybe there are already some, but I will not go looking for them as the curriculum is beefy enough). 

I'm against putting it in LG 2-6, it's not a principle.

I'm not happy with a new LG 2-10. Currently our LGs are a - not perfect but sufficient - topic tree. If we add sub-topics as new LGs, we clutter this topic tree and also elevate the perceived importance of that sub-topic (sounds weird, but that's how new trainers might perceive it according to my experience).

So I'm for option 1, but keep it very brief and make it R3.

**Comment by gernotstarke** [comment]
I go with @alxlo arguments - so let's add a bullet point to 2-4 ("deployment concepts"), make it R3 (so it is completely optional).

**Comment by gernotstarke** [comment]
@mahboubagharbi please have a look at #260, and propose stuff we might remove...

**Comment by gernotstarke** [comment]
The majority of votes were in favor of adding a new LG 2-10 with R3 and include a few deployment principles.

**Comment by gernotstarke** [comment]
I checked all the sources given by @rhoadesre above, but none serves as a stable, long-lasting reference (Michael Nygards' book does, but he does not collect deployment principles in it). I'm unsure wether the old book by Len Bass (DevOps: A Software Architect's Perspective) is good and current enough for us.

I condensed the above proposals and statements to the following, and hope I translated to acceptable German:

```
==== LG 2-10: Deployment (R3)

Software architects:

* know that software deployment is the process of making new or updated software available to its users
* are able to name and explain fundamental concepts of software deployment, for example:
** automate deployments
** repeatable builds
** consistent environments (e.g. use immutable and disposable infrastructure)
** version-control everything
** releases are easy-to-undo

```

see #324

**Comment by programming-wolf** [comment]
Closed with #324
